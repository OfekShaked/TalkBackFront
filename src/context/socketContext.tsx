import {io} from "socket.io-client";
import React from 'react';
import config from '../configs/config';
const SOCKET_URL = config.apiUrl;

export const socket = io(SOCKET_URL, { transports: ['websocket', 'polling', 'flashsocket'] });
export const SocketContext = React.createContext<any>(null);
