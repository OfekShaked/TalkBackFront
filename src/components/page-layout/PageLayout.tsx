import React, { useState, useEffect, useContext } from 'react';
import clsx from 'clsx';
import { useTheme } from '@material-ui/core/styles';
import CssBaseline from '@material-ui/core/CssBaseline';
import useStyles from './PageLayoutStyle';
import Header from '../header/Header';
import SignUp from '../../views/sign-up/SignUp';
import Login from '../../views/log-in/Login';
import Logout from '../log-out/Logout';
import ContactScreen from '../../views/contact-screen/ContactScreen';
import WebsiteDown from '../../views/website-down/WebsiteDown';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import { isAuthorized } from '../../services/auth.service'
import { SocketContext } from '../../context/socketContext'
import { handleError } from '../../services/errorHandling.service'

export default function PageLayout() {
  const [currentPage, setCurrentPage] = useState("");
  const socket = useContext(SocketContext);
  const theme = useTheme();
  const classes = useStyles(theme);
  const [open, setOpen] = useState(false);
  const [isLoggedIn, setIsLoggedIn] = useState<Boolean>(false);
  const [isServerOnline,setIsServerOnline] = useState(false);

  const handleDrawerOpen = () => {
    setOpen(true);
  };

  const handleDrawerClose = () => {
    setOpen(false);
  };

  useEffect(() => {
    try {      
      const getAuth = async () => {
        setIsLoggedIn(await isAuthorized());
      }
      getAuth();
    } catch (err) {
      handleError(err);
    }
  }, [])

  useEffect(() => {
    try {
      if (isLoggedIn) {
        socket.emit("user_online", localStorage.getItem("username"));
        setCurrentPage("/contact");
      }
      else {
        socket.emit("logout");
        setCurrentPage("/signin");
      }
    } catch (err) {
      handleError(err);
    }
  }, [isLoggedIn])

  useEffect(()=>{
    setIsServerOnline(socket.connected);
    socket.on("connect",()=>{
      setIsServerOnline(socket.connected);      
    })
  },[socket])


  return (
    <div className={classes.root}>
      <CssBaseline />
      <BrowserRouter>
      <Switch>
        <Header handleDrawerOpen={handleDrawerOpen} open={open} handleDrawerClose={handleDrawerClose} isLoggedIn={isLoggedIn}></Header>

        <main
          className={clsx(classes.content, {
            [classes.contentShift]: open,
          })}
        >
          <div className={classes.drawerHeader} />
          {isServerOnline?<>
          <Route path="/signin" component={() => isLoggedIn ? <ContactScreen /> : <Login setIsLoggedIn={setIsLoggedIn} />} />
          <Route path="/register" component={() => isLoggedIn ? <ContactScreen /> : <SignUp/>}/>
          <Route path="/contact" component={ContactScreen} />
          <Route path="/logout" component={() => <Logout setIsLoggedIn={setIsLoggedIn} />} />
          <Route path="*" component={() => isLoggedIn ? <ContactScreen /> : <Login setIsLoggedIn={setIsLoggedIn} />} />
          </>:
          <Route component={() =><WebsiteDown/>} />
          }
        </main>
        </Switch>
      </BrowserRouter>
    </div>
  );
}