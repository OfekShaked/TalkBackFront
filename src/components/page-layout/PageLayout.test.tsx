import React from 'react';
import renderer from 'react-test-renderer';
import PageLayout from './PageLayout';
import { SocketContext } from '../../context/socketContext';
import SocketMock from 'socket.io-mock';
import { isAuthorized } from '../../services/auth.service'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { mount, configure } from 'enzyme';
import {act} from 'react-dom/test-utils';

configure({adapter: new Adapter()});

jest.mock('../../services/auth.service', () => ({
      isAuthorized: jest.fn(),
      getCurrentUser: jest.fn(()=>{return "ofek"})
}));


let socket = new SocketMock()

describe("page layout component", () => {
    it('renders PageLayout component', async() => {
        (isAuthorized as jest.Mock).mockImplementation(()=>{return false;})
        const component = renderer.create(
            <PageLayout></PageLayout>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('is not authorized', async () => {
        (isAuthorized as jest.Mock).mockImplementation(()=>{return false;})
        const component = await mount(
            <SocketContext.Provider value={socket.socketClient}>
                <PageLayout></PageLayout>
            </SocketContext.Provider>
        );
        expect(component.text()).toContain("Sign In")
    });

    it('is authorized', async () => {
        (isAuthorized as jest.Mock).mockImplementation(()=>{return true;})
        let component:any ;
        await act(async ()=>{
            component = await mount(
            <SocketContext.Provider value={socket.socketClient}>
                <PageLayout></PageLayout>
            </SocketContext.Provider>
        )
        });
        if (!component) {
            throw new Error("User is null");
          }
        expect(component).toBeDefined()
        expect(component.text()).toContain("Status")
        
    });

})

