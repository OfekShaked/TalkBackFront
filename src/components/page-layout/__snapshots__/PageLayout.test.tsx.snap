// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page layout component renders PageLayout component 1`] = `
<div
  className="makeStyles-root-1"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-5 mui-fixed MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <button
        aria-label="open drawer"
        className="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-7 MuiIconButton-colorInherit MuiIconButton-edgeStart"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
      <h6
        className="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap"
      >
        What's On My App
      </h6>
    </div>
  </header>
  <div
    className="MuiDrawer-root MuiDrawer-docked makeStyles-drawer-9"
  >
    <div
      className="MuiPaper-root MuiDrawer-paper makeStyles-drawerPaper-10 MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft MuiPaper-elevation0"
      style={
        Object {
          "visibility": "hidden",
        }
      }
    >
      <div
        className="makeStyles-drawerHeader-11"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </div>
      <hr
        className="MuiDivider-root"
      />
      <ul
        className="MuiList-root MuiList-padding"
      >
        <a
          aria-current="page"
          aria-disabled={false}
          className="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button active"
          href="/signin"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          style={Object {}}
          tabIndex={0}
        >
          <div
            className="MuiListItemIcon-root"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
          </div>
          <div
            className="MuiListItemText-root"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              Sign-In
            </span>
          </div>
          <span
            className="MuiTouchRipple-root"
          />
        </a>
        <a
          aria-current={null}
          aria-disabled={false}
          className="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
          href="/register"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemIcon-root"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </div>
          <div
            className="MuiListItemText-root"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              Register
            </span>
          </div>
          <span
            className="MuiTouchRipple-root"
          />
        </a>
      </ul>
    </div>
  </div>
  <main
    className="makeStyles-content-3"
  >
    <div
      className="makeStyles-drawerHeader-2"
    />
    <main
      className="MuiContainer-root MuiContainer-maxWidthXs"
    >
      <div
        className="makeStyles-paper-12"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular makeStyles-avatar-13 MuiAvatar-colorDefault"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
            />
          </svg>
        </div>
        <h1
          className="MuiTypography-root MuiTypography-h5"
        >
          Sign In
        </h1>
        <form
          className="makeStyles-form-14"
          method="post"
          noValidate={true}
          onSubmit={[Function]}
        >
          <div
            className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
          >
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <div
                className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                  data-shrink={false}
                  htmlFor="username"
                  id="username-label"
                >
                  Username
                  <span
                    aria-hidden={true}
                    className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoComplete="string"
                    autoFocus={false}
                    className="MuiInputBase-input MuiOutlinedInput-input"
                    disabled={false}
                    id="username"
                    name="username"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={true}
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden={true}
                    className="PrivateNotchedOutline-root-16 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      className="PrivateNotchedOutline-legendLabelled-18"
                    >
                      <span>
                        Username
                         *
                      </span>
                    </legend>
                  </fieldset>
                </div>
                
              </div>
            </div>
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <div
                className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                  data-shrink={false}
                  htmlFor="password"
                  id="password-label"
                >
                  Password
                  <span
                    aria-hidden={true}
                    className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoComplete="current-password"
                    autoFocus={false}
                    className="MuiInputBase-input MuiOutlinedInput-input"
                    disabled={false}
                    id="password"
                    name="password"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={true}
                    type="password"
                    value=""
                  />
                  <fieldset
                    aria-hidden={true}
                    className="PrivateNotchedOutline-root-16 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      className="PrivateNotchedOutline-legendLabelled-18"
                    >
                      <span>
                        Password
                         *
                      </span>
                    </legend>
                  </fieldset>
                </div>
                
              </div>
            </div>
          </div>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-15 MuiButton-containedPrimary MuiButton-fullWidth"
            disabled={false}
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="submit"
          >
            <span
              className="MuiButton-label"
            >
              Log In
            </span>
          </button>
        </form>
      </div>
    </main>
  </main>
</div>
`;
